Assets/MapGenerator.cs

```diff
@@ -0,0 +1,274 @@
+	[Range(0, 8)]
+	public int DeathLimit = 5;		//if the number of walls around the selected non-blank point is not too high then kill it
+
+
+	int[,] map;
+	int[,] newMap;
+	//MapCol[] map;
+	int[] Topography;
+
+	void Start() 
+	{
+		GenerateMap();
+	}
+
+	void Update() 
+	{
+		if (Input.GetMouseButtonDown(0)) 
+		{
+			GenerateMap();
+		}
+	}
+
+	void GenerateMap() 
+	{
+		height = (int)(groundDepth + hillDepth + skyDepth);
+		//map = new MapCol[width];
+		map = new int[width,height];
+		Topography = new int[width];
+		heightSum = 0;
+		ClearMap ();
+		//RandomFillMap();
+		GenerateHills ();
+		SeedCaves ();
+
+		for (int i = 0; i < SmoothingIterations; i ++) 
+		{
+			SmoothMap();
+		}
+	}
+
+	void GenerateHills()
+	{
+
+		if (useRandomSeed) 
+		{
+			seed = Time.time.ToString();
+		}
+		for(int x = 0;x < width;x++)
+		{
+			float xf = (float)x / (float)(width-1);
+			//height will evaluate to a value ~ between 0 and 1
+			GroundTop = contourCurve.Evaluate(xf) * heightScale * Mathf.PerlinNoise(xf * xScale, yNoise);
+			//multiply by hillDepth to get a realistic range of hills
+			GroundTop *= hillDepth;
+			//Add a solid chunk of ground underneath the hills
+			GroundTop += groundDepth;
+			//map[x].Topography = (int)GroundTop;
+			Topography[x] = (int)GroundTop;
+			//heightSum += map[x].Topography;
+			heightSum += Topography[x];
+			for(int y = 0;y < height;y++)
+			{
+				if(y <= GroundTop)
+				{
+					//map[x].Column[y] = 1;
+					map[x,y] = 1;
+				}
+				else
+				{
+					y = (int)height; 
+				}
+			}
+		}
+	}
+
+	void ClearMap()
+	{
+		for (int x = 0; x < width; x ++) 
+		{
+			for (int y = 0; y < height; y ++) 
+			{
+				map[x,y] = 0;
+				//map[x].Column[y] = 0;
+			}
+		}
+	}
+
+	void SwapMaps()
+	{
+		for (int x = 0; x < width; x ++) 
+		{
+			for (int y = 0; y < height; y ++) 
+			{
+				map[x,y] = newMap[x,y];
+			}
+		}
+	}
+
+	void SeedCaves()	//generate a number of 'holes' in the ground where caves could form later
+	{
+		if (useRandomSeed) 
+		{
+			seed = Time.time.ToString();
+		}
+		System.Random pseudoRandom = new System.Random(seed.GetHashCode());
+		for (int x = 1; x < width-1; x ++)  //Don't seed the edges
+		{
+			//for (int y = 0; y < map[x].Topography-1; y ++) //stops 1 block short of ground surface
+			for (int y = 3; y < Topography[x]-5; y ++) //stops 5 block short of ground surface
+			{
+				//if(map[x].Column[y] != 0)
+				if(map[x,y] != 0)
+				{
+					//should we seed a cave here?
+					if(pseudoRandom.Next(0,100) < randomCaveSeedPercent)
+					{
+						//map[x].Column[y] = 0;
+						map[x,y] = 0;
+					}
+				}
+			}
+		}
+	}
+
+
+	void SmoothMap() 
+	{
+		newMap = new int[width,height];
+		for (int x = 0; x < width; x ++) 
+		{
+			for (int y = 0; y < height; y ++) 
+			{
+				int neighbourWallTiles = GetSurroundingWallCount(x,y);
+				if(x > 0 && x  < width -1 && y > 0 && y < Topography[x] - 5)  //do not process any cells that are out of bounds
```

Comment:
Perfect example of wanting a constant, since this number was used in the previous method